tests/counterfactual/test_handlers_explanation.py

```diff
@@ -0,0 +1,197 @@
+            prob_bottle_shatters_if_sally,
+            torch.tensor(0.0),
+        ),
+    )
+
+    bottle_shatters = pyro.sample("bottle_shatters", dist.Bernoulli(new_bsp))
+
+    return {
+        "sally_throws": sally_throws,
+        "bill_throws": bill_throws,
+        "sally_hits": sally_hits,
+        "bill_hits": bill_hits,
+        "bottle_shatters": bottle_shatters,
+    }
+
+
+def test_single_layer_stones():
+    observations = {
+        "prob_sally_throws": 1.0,
+        "prob_bill_throws": 1.0,
+        "prob_sally_hits": 1.0,
+        "prob_bill_hits": 1.0,
+        "prob_bottle_shatters_if_sally": 1.0,
+        "prob_bottle_shatters_if_bill": 1.0,
+    }
+
+    with MultiWorldCounterfactual() as mwc:
+        with PartOfCause({"sally_throws": 0.0}, prefix="preempt_"):
+            with condition(
+                data={k: torch.as_tensor(v) for k, v in observations.items()}
+            ):
+                with pyro.poutine.trace() as trace:
+                    stones_bayesian_model()
+
+    tr = trace.trace.nodes
+
+    with mwc:
+        preempt_sally_throws = gather(
+            tr["preempt_sally_throws"]["value"],
+            IndexSet(**{"sally_throws": {0}}),
+            event_dim=0,
+        )
+        int_sally_hits = gather(
+            tr["sally_hits"]["value"], IndexSet(**{"sally_throws": {1}}), event_dim=0
+        )
+
+        obs_bill_hits = gather(
+            tr["bill_hits"]["value"], IndexSet(**{"sally_throws": {0}}), event_dim=0
+        )
+        int_bill_hits = gather(
+            tr["bill_hits"]["value"], IndexSet(**{"sally_throws": {1}}), event_dim=0
+        )
+        int_bottle_shatters = gather(
+            tr["bottle_shatters"]["value"],
+            IndexSet(**{"sally_throws": {1}}),
+            event_dim=0,
+        )
+
+    outcome = {
+        "preempt_sally_throws": preempt_sally_throws.item(),
+        "int_sally_hits": int_sally_hits.item(),
+        "obs_bill_hits": obs_bill_hits.item(),
+        "int_bill_hits": int_bill_hits.item(),
+        "intervened_bottle_shatters": int_bottle_shatters.item(),
+    }
+
+    assert list(outcome.values()) == [0, 0.0, 0.0, 1.0, 1.0] or list(
+        outcome.values()
+    ) == [1, 1.0, 0.0, 0.0, 1.0]
+
+
+@pytest.mark.parametrize("antecedents", [None, ("sally_throws",), ("bill_hits",)])
+def test_two_layer_stones(antecedents):
+    observations = {
+        "prob_sally_throws": 1.0,
+        "prob_bill_throws": 1.0,
+        "prob_sally_hits": 1.0,
+        "prob_bill_hits": 1.0,
+        "prob_bottle_shatters_if_sally": 1.0,
+        "prob_bottle_shatters_if_bill": 1.0,
+    }
+
+    evaluated_node_counterfactual = {"sally_throws": 0.0}
+    witness_preemptions = {"bill_hits": factual_preemption(antecedents=antecedents)}
+
+    pinned_preemption_variables = {
+        "preempt_sally_throws": torch.tensor(0),
+        "witness_preempt_bill_hits": torch.tensor(1),
+    }
+
+    with MultiWorldCounterfactual() as mwc, PartOfCause(
```

Comment:
The default behavior of `black` for nested context managers is kind of gross... At some point we should find a way to reconfigure it to produce more readable code.